[DOC] data container conventions, conversions, and data sets documentation

fkiraly opened the pull request:
DRAFT pull request on data container conventions and related functionality.

This is based on, and branching off the state of #1225 - and requests of @mloning and @TonyBagnall to document this.

The increment in this PR is the notebook `examples/00_datatypes` which is supposed to explain:
* the different core in-memory data container datatypes for series and panels used in `sktime`, with examples to play around with
* validity checking, conversion, register, and type functionality
* possibly: data sets loading functionality

Still work in progress, but would already appreciate a look at section 1.

TonyBagnall reviewed (commented):
I would like a clear list at the top of the abstract data types and implementations allowed for each, preferably also a diagram

fkiraly commented:
Ready for review now.

Lovkush-A commented on sktime/datatypes/_convert.py:
```diff
@@ -66,12 +65,17 @@
+import numpy as np
+import pandas as pd
 
 from sktime.datatypes._series import convert_dict_Series
 from sktime.datatypes._series import infer_mtype_dict_Series
 
 from sktime.datatypes._panel import convert_dict_Panel
```
Do we not have `infer_mtype_dict_Panel` because there is only one mtype for Panel data?

fkiraly replied:
Actually, I'm looking to deprecate the `infer_mtype_dict`-s, because there is no 1:1 mapping onto types.
Perhaps I should have simply deleted this.

Instead, it will be handled by the `mtype` function which loops through the checks.
In all the checks (in `check_is`), the first line is a type check, so the "first level of checks" is equivalent to using the dictionary.

fkiraly replied:
more precisely, there are two `mtype` functions: the old one, in `_convert`, and a new one, in `_check`.
I want to deprecate the one in `_convert` (currently only does it for `Series`) and replace it by the one in `_check` (this one is generic and relies on full checks rather than dictionaries).

Lovkush-A replied:
might be worth adding a todo item or comment saying that the dicts are to be deprecated/deleted

fkiraly replied:
I decided to removed the old `mtyp` function and replaced it by the new one in the module export.

Since everything was referring to the module export rather than the precise location, and both functions had the same signature, nothing broke - so no deprecation is actually needed, and the problem should be fixed?

Lovkush-A reviewed (commented):
First batch of comments. I have only looked at the smaller files/changes. From what I have seen so far, things look good and well organised.

Note for future: it would be easier to review this if you did the minor docstring changes in a separate PR.

fkiraly commented on sktime/datatypes/tests/test_convert.py:
```diff
@@ -0,0 +1,85 @@
+    fixtures = dict()
+
+    for mtype in mtypes:
+        # if we don't do this we get into a clash between linters
+        mtype_long_variable_name_to_avoid_linter_clash = mtype
```
not sure what exactly is happening, but it seems the linters fight each other if you are between 80 and 90 characters or so.

fkiraly commented on sktime/datatypes/__init__.py:
```diff
@@ -1,5 +1,29 @@
 # -*- coding: utf-8 -*-
 """Module exports: series type converters."""
```
yes, fixed.

fkiraly commented on sktime/datatypes/_convert.py:
```diff
@@ -132,6 +137,15 @@ def convert(obj, from_type: str, to_type: str, as_scitype: str, store=None):
     if obj is None:
         return None
 
+    if not isinstance(to_type, str):
+        raise TypeError("to_type must be a str or list of str")
```
fixed this

Lovkush-A reviewed (commented):
Looked at notebook. This will help when I look at the other files!

* I dont think this will be final version of notebook, but it is good for the purposes of this PR.
* Some things I suggest changing in future iterations of this notebook:
  * Not having it as the 00 notebook. The notebooks in which we actually do interesting stuff, like forecasting and classification, should come first, and this notebook should be later.
  * Adding contents at the top, for the sections, and subsections, and subsubsections. I presume there is a way to do this automatically.
  * Reorganising the notebook and/or making the content on types more accessible. Currently, the first section on types is dense and abstract, and I imagine difficult to follow for most users. There are various ideas I have for this (e.g. giving example of each machine type first, before the full specifications), but can be discussed in separate pr.

fkiraly commented:
>I dont think this will be final version of notebook, but it is good for the purposes of this PR.

Well, better *this* notebook than *no notebook whatsoever*, no?
This should help our users who were previously struggling with the panel data types because the expected format was nowhere properly recorded...

fkiraly commented on examples/00_datatypes_and_datasets.ipynb:
```diff
@@ -0,0 +1,1084 @@
+   "source": [
+    "from sktime.datatypes import check_raise\n",
+    "\n",
+    "# uncomment this to see the informative error message\n",
+    "#\n",
+    "# check_raise(X, mtype=\"pd-multiindex\")"
```
no, because then it will raise an informative error message, which will then be caught by CI `test_examples`, which thinks that the notebooks breaks (as opposed to breaking in a place where it's supposed to).

unless you have a better idea how to handle this, of course

Lovkush-A replied:
Oh, I see. At my work, I saw somebody getting around this with a try except block, and in the except block you can print the error message somehow. I will try to find it

fkiraly replied:
hmmmm - but wouldn't this confuse notebook users since they think the try/except is part of it?

Lovkush-A replied:
Yes, try/except could confuse users... looks like there is no ideal option. The issue with current setup is that things that require uncommenting tend to be optional in tutorials I have seen.

Maybe have the instruction to uncomment the cell be a part of the markdown, rather than a comment itself?

fkiraly replied:
yes, agreed, text in comments is more likely to be ignored, so I moved the instruction to the markdown where it will be percieved as part of the "voiceover"

Lovkush-A commented:
> Well, better _this_ notebook than _no notebook whatsoever_, no?

Yup, definitely! Not only good for end-users, but developers too.

fkiraly commented:
>This should help our users who were previously struggling with the panel data types because the expected format was nowhere properly recorded...

What's even more funny, after I did all due diligence here, there are *seven* panel data (m)types in use throughout the package...

fkiraly commented:
>Not having it as the 00 notebook. The notebooks in which we actually do interesting stuff, like forecasting and classification, should come first, and this notebook should be later.

Hm, how would you organise it then? An "appendix notebook" where numbering starts with A, B, C, etc, and which is linked from the "interesting stuff" notebooks?

fkiraly commented:
>Adding contents at the top, for the sections, and subsections, and subsubsections. I presume there is a way to do this automatically.

Sadly, I think, no 😢 

There is a semi-automatic way, which is putting it out as an issue waiting for helpful contributors such as @bilal-196 in #1200 to pick it up (thanks again!) 😄 while I work on creating notebook content, sort of a division of labour...

Would be great if this could be somehow automated though. Perhaps we could put *that* out as an issue?

fkiraly commented:
>Reorganising the notebook and/or making the content on types more accessible. Currently, the first section on types is dense and abstract, and I imagine difficult to follow for most users. There are various ideas I have for this (e.g. giving example of each machine type first, before the full specifications), but can be discussed in separate pr.

Looking forward to that 😄 feel free to pick it up once this is in

My main motivation for doing this was sorting this out so I can work on the transformers, I think it's a pre-condition to have the panel types, checks, and conversions sorted to do the refactor cleanly

fkiraly commented on examples/00_datatypes_and_datasets.ipynb:
```diff
@@ -0,0 +1,1084 @@
+   "source": [
+    "from sktime.datasets import load_arrow_head\n",
+    "\n",
+    "X, y = load_arrow_head(return_X_y=True)"
```
I think we can assume here it's going to be fixed as proposed?
For the didactic purpose of the notebook, it should load the entire data set.

Lovkush-A replied:
resolved. agree not to make suggested change.

Lovkush-A commented:
> Hm, how would you organise it then? An "appendix notebook" where numbering starts with A, B, C, etc, and which is linked from the "interesting stuff" notebooks?

I think a short-term option is to renumber it from 00 to 05. Better options depend a lot on what we consider to be the 'ideal end state' of the documentation, which is unknown to me. This stuff on data conversion feels more appropriate for a 'how-to' section of documentation rather than 'tutorial' section.

fkiraly commented:
>Better options depend a lot on what we consider to be the 'ideal end state' of the documentation, which is unknown to me. This stuff on data conversion feels more appropriate for a 'how-to' section of documentation rather than 'tutorial' section.

How about this for a practical solution: we have "tutorials" with interesting stuff, numbered 1, 2, etc. 
Then, we have "appendix" type tutorials, into more dense/technical parts of the toolbox and its infrastructure, numbered A, B, etc. They would be linked from the tutorials, e.g., "for loading your own data, consult notebook A on data types and data validation".

Lovkush-A commented:
> How about this for a practical solution: we have "tutorials" with interesting stuff, numbered 1, 2, etc.
> Then, we have "appendix" type tutorials, into more dense/technical parts of the toolbox and its infrastructure, numbered A, B, etc. They would be linked from the tutorials, e.g., "for loading your own data, consult notebook A on data types and data validation".

Certainly sounds like a sensible option. But I hesitate to say anything definitive because the decision of how to organise the notebook examples folder depends on the whole vision of the documentation for the package (https://github.com/alan-turing-institute/sktime/issues/1059). E.g. maybe the better solution is two have two separate folders for notebooks; one for beginner/intro tutorials and the other for advanced/how-to tutorials. Or something else entirely.

fkiraly commented:
>Certainly sounds like a sensible option. But I hesitate to say anything definitive because the decision of how to organise the notebook examples folder depends on the whole vision of the documentation for the package (#1059). E.g. maybe the better solution is to have two separate folders for notebooks; one for beginner/intro tutorials and the other for advanced/how-to tutorials. Or something else entirely.

My advice would be, let's solve one problem at a time - let's solve the small problem and avoid getting stalled by trying to open up a scope which is far beyond this PR.

I don't think a complete vision for documentation is stated currently; it's also easier to develop if we have "all the parts" ready. Next week's doc sprint will be a good time to creat such a vision, but I would not make this Pr dependent on this - I feel there will be multiple iterations of discussions for the duration of a month or so, just an educated guess.

Lovkush-A commented:
> but I would not make this Pr dependent on this 

i agree. my original comment did say the notebook was good for this pr and my suggestions were for future iterations of the notbeook.

fkiraly commented:
>i agree. my original comment did say the notebook was good for this pr and my suggestions were for future iterations of the notbeook.

Yes, let's revisit the "target state" of documentation during the doc sprint!

fkiraly commented:
so, regarding the `mtype` function: I removed the old one and replaced it by the new one in the module export.

Since everything was referring to the module export rather than the precise location, nothing broke - so no deprecation is actually needed.

fkiraly commented on examples/00_datatypes_and_datasets.ipynb:
```diff
@@ -0,0 +1,1084 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Set-up instructions:** this notebook give a tutorial on the forecasting learning task supported by `sktime`.\n",
```
yes, updated this.

fkiraly commented:
?? why is readtheocs failing now?

Lovkush-A reviewed (commented):
looked at _check.py.  only minor comments and suggestions. some are suggestions I think that can be made in this PR, and a couple of others that I think are better in a later iteration/tidying of these files.

question: at this rate, it is going to take a while for me to review the pr. is there anything in particular you want me to check (or not check) so that we can get this PR reviewed more quickly?

fkiraly commented:
>is there anything in particular you want me to check (or not check) so that we can get this PR reviewed more quickly?

I would say, focus on the general logic/framework as opposed to the individual checks or converters - many of those were moved into the package rather than implemented from scratch.

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+    is_mtype = np.array(is_mtype)
+    res = [mtypes[i] for i in range(len(mtypes)) if is_mtype[i]]
+
+    if np.sum(is_mtype) > 1:
+        raise TypeError(f"Error in check_is, more than one mtype identified: {res}")
```
yes, it should not happen but could happen by accident, so it's worth that the error tells you that.

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+# -*- coding: utf-8 -*-
+"""Machine type checkers for scitypes.
+
+Exports
+-------
+check_is(obj, mtype: str, scitype: str)
+    checks whether obj is mtype for scitype
+    returns boolean yes/no and metadata
+
+check_raise(obj, mtype:str, scitype:str)
+    checks whether obj is mtype for scitype
+    returns True if passes, otherwise raises error
+"""
```
done

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+
+    mtypes = np.array([x[0] for x in list(check_dict.keys()) if x[1] == as_scitype])
+
+    is_mtype = [check_is(obj, mtype=mtype, scitype=as_scitype) for mtype in mtypes]
+    is_mtype = np.array(is_mtype)
+    res = [mtypes[i] for i in range(len(mtypes)) if is_mtype[i]]
```
yes, that works too - I'm just generally paranoid about hitting the linter max character bound

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+    if as_scitype not in valid_as_scitypes:
+        raise TypeError(as_scitype + " is not a supported scitype")
+
+    mtypes = np.array([x[0] for x in list(check_dict.keys()) if x[1] == as_scitype])
```
fixed

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+    if obj is None:
+        return None
+
+    valid_as_scitypes = list(set([x[1] for x in list(check_dict.keys())]))
```
fixed

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+
+    Parameters
+    ----------
+    obj : object to convert - any type, should comply with mtype spec for as_scitype
+    as_scitype : str - name of scitype the object "obj" is considered as
+
+    Returns
+    -------
+    str - the type to convert "obj" to, a valid mtype string
+        or None, if obj is None
```
fixed

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+
+
+def check_is(
+    obj, mtype: str, scitype: str = None, return_metadata=False, var_name="obj"
```
that would make it extremely complex and hard to read (`Union[List[str],str]`?), perhaps better not?

Lovkush-A replied:
Sometimes the truth is ugly...

I think correctness has a higher priority than aesthetics when it comes to type hints.

fkiraly replied:
alright

Lovkush-A replied:
I don't think you changed the type hints.

Lovkush-A replied:
OK, I'm confused. I just checked again and now see it has been updated. I don't know what is going on with GitHub interface.

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+        else:
+            check_passed = res
+
+        if check_passed:
+            return res
```
hm, why? if you want that information, you would run `mtype`.
Or do you mean, there is a situation where you want both "which mtype" and the metadata?

In addition, would you want it returned as part of the metadata?

Lovkush-A replied:
If the function `check_is` is only used internally, then my suggestion would not be necessary.

fkiraly replied:
well, you still would use `mtype` for that use case, it returns the `mtype`.

Compare `isinstance`, note that it also does not return the type that the object is - to read the type, you wouuld use `type`.
In terms of signature, `check_is` and `mtype` mirror `isinstance` and `mtype`.

fkiraly replied:
resolved, since you say it's fine to discuss in future iterations.

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+        if True, returns all three return objects
+    var_name: str, optional, default="obj" - name of input in error messages
+
+    Returns
+    -------
+    valid: bool - whether obj is a valid object of mtype/scitype
+    msg: str or list of str - error messages if object is not valid, otherwise None
+            str if mtype is str; list of len(mtype) with message per mtype if list
+            returned only if return_metadata is True
+    metadata: dict - metadata about obj if valid, otherwise None
+            returned only if return_metadata is True
+        fields:
+            "is_univariate": bool, True iff series has one variable
+            "is_equally_spaced": bool, True iff series index is equally spaced
```
the refactor would then just set `return_metadata` = True, no?
I think it might be a bit cumbersome for the user if there are three return arguments?

There's two usage scenarios, one for users who want to check their input data, and one for internal use. I thought the metadata would only be useful in the second, so default behaviour (for the user) should be the "straigtforward expected return type".

Lovkush-A replied:
> the refactor would then just set `return_metadata` = True, no?

that would be the quick refactor. However, you would also be able to tidy up many functions and improve readability of the code, e.g. remove need to define the `ret` function in various functions.

> There's two usage scenarios, one for users who want to check their input data, and one for internal use. I thought the metadata would only be useful in the second, so default behaviour (for the user) should be the "straigtforward expected return type".

I cannot think of anything on the top of my head for how to deal with this. Hence why I think the current state is good for now, and we can refactor later.

fkiraly commented:
@Lovkush-A, are you still reviewing this?

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+from sktime.datatypes._series import check_dict_Series
+from sktime.datatypes._registry import mtype_to_scitype
+
+# pool convert_dict-s and infer_mtype_dict-s
```
done, fixed

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+        raise TypeError(res[1])
+
+
+def mtype(obj, as_scitype: str = None):
```
better not in this PR, because `mtype` is called in a few external places already - if we decide to rename, there should be a dedicated PR

Lovkush-A commented:
> @Lovkush-A, are you still reviewing this?

I intend to finish reviewing this weekend, i.e. later today.

Lovkush-A commented on sktime/datatypes/_panel/_check.py:
```diff
@@ -0,0 +1,253 @@
+    bad_inds = [i for i in range(n) if not check_res[i][0]]
+
+    if len(bad_inds) > 0:
+        msg = f"{var_name}[i] must be Sries of mtype pd.DataFrame, not at i={bad_inds}"
```
```suggestion
        msg = f"{var_name}[i] must be Series of mtype pd.DataFrame, not at i={bad_inds}"
```

fkiraly replied:
fixed

Lovkush-A commented on sktime/datatypes/_series/_check.py:
```diff
@@ -0,0 +1,208 @@
+check_dict = dict()
+
+
+def check_pdDataFrame_Series(obj, return_metadata=False, var_name="obj"):
+
```
missing docstring

Lovkush-A replied:
same for the other functions in this file

fkiraly replied:
these are private (so no docstring needed by linter), also all functions have the same signature which is detailed in the file docstring to avoid multiplication

Lovkush-A commented on sktime/datatypes/tests/test_convert.py:
```diff
@@ -0,0 +1,85 @@
+SCITYPES = [sci[0] for sci in SCITYPE_REGISTER]
+
+
+@pytest.mark.parametrize("scitype", SCITYPES)
+def test_convert(scitype):
+    """Tests that conversions for scitype agree with from/to example fixtures.
```
I think testing logic here is correct, but is hard to follow.  E.g.
* This line `n_fixtures = len(fixtures[mtypes[0]])`only uses the 0 index of `mtypes`. It is not clear upfront that this will not break things, or why this is done.
* In my head, I'd have expected the logic to look something like this:
```
for from_mtype in conv_mat.index:
for to_mtype in conv_mat.columns:
check conversion is defined, if not pass.
loop through examples, do conversions and checks.
```
```suggestion
@pytest.mark.parametrize("scitype", SCITYPES)
def test_convert(scitype):
    """Tests that conversions for scitype agree with from/to example fixtures.
```
```suggestion
@pytest.mark.parametrize("scitype", SCITYPES)
def test_convert(scitype):
    """Tests that conversions for scitype agree with from/to example fixtures.
```

fkiraly replied:
>In my head, I'd have expected the logic to look something like this:

Isn't that precisely what happens?

fkiraly replied:
there's some additional checking, since some fixtures are "not defined", e.g., if you try to convert a strictly multivariate series to a univariate one - `None` then indicates that it's not possible. The `None` related edge cases need checking.

Lovkush-A replied:
> Isn't that precisely what happens?

Yes, but it is not obvious that is what is happening:
* e.g. the outermost loop ranges through `range(n_fixtures)` whereas I would expect outermost loop to be for columns and indices of conversion matrix
* you loop through rows and indices of conversion matrix indirectly via the variable `mtypes`, whereas I think it would be easier to read if you looped through the indices and columns directly.

Here is something that I think would be clearer:

```
    for from_type in conv_mat.index:
        for to_type in conv_mat.columns:
            for to_fixture, from_fixture in zip(fixtures[to_type], fixtures[from_type])
```

Anyway, for purposes of this review, it is all good as is. If you want to refactor a little bit, you can, otherwise, leave as is.

Lovkush-A reviewed (changes requested):
I have now skimmed through all files and logic and changes make sense. Structure of files is easy to follow. In particular, the tests look like they check everything we want.

All comments made here are minor (e.g. typos, missing docstrings). There are also some unresolved conversations from my previous reviews.

I'll approve when all conversations have been appropriately resolved!

fkiraly commented on sktime/datatypes/_examples.py:
```diff
@@ -17,44 +17,53 @@
 
-def get_example(mtype: str, as_scitype: str, ind: int = 0, return_lossy: bool = False):
+def get_examples(mtype: str, as_scitype: str, return_lossy: bool = False):
     """Retrieve the ind-th example for mtype mtype, scitype as_scitype.
```
updated

Lovkush-A replied:
seen and resolved.

fkiraly commented on sktime/datatypes/_examples.py:
```diff
@@ -17,44 +17,53 @@
-    if return_lossy=True, also returns:
-    lossy: bool - whether the example is a lossy representation
+    fixtures: dict with integer keys, elements being
+        fixture - example for mtype mtype, scitype as_scitype
```
updated

fkiraly commented on sktime/datatypes/_registry.py:
```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+
```
yes, or the powers that lint will be displeased

Lovkush-A replied:
I think you forgot to do this.

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+    ValueError if mtype input argument is not of expected type
+    """
+    obj_long_name_for_avoiding_linter_clash = obj
+    res = check_is(
+        obj=obj_long_name_for_avoiding_linter_clash,
+        mtype=mtype,
+        scitype=scitype,
+        return_metadata=True,
+        var_name=var_name,
+    )
+
+    if res[0]:
+        return True
+    else:
+        raise TypeError(res[1])
```
changed

fkiraly commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,187 @@
+    ValueError if mtype input argument is not of expected type
+    """
+    obj_long_name_for_avoiding_linter_clash = obj
+    res = check_is(
+        obj=obj_long_name_for_avoiding_linter_clash,
+        mtype=mtype,
+        scitype=scitype,
+        return_metadata=True,
+        var_name=var_name,
+    )
+
+    if res[0]:
+        return True
+    else:
+        raise TypeError(res[1])
```
yes, it just converts it to string and prints the list as a list like `['message1', 'message2']` which should be self-explanatory?

Lovkush-A replied:
I don't know! Have to see what users say I guess.

Lovkush-A commented on sktime/datatypes/_check.py:
```diff
@@ -0,0 +1,200 @@
+
+def check_is(
+    obj,
+    mtype: str,
+    scitype: Union[str, list] = None,
```
```suggestion
    mtype: Union[str, list[str]],
    scitype: str = None,
```
* I think type hints were mixed up between mtype and scitype
* I think you should be specific that it is a list of string, not just any list

Lovkush-A reviewed (changes requested):
tiny change suggestion. mixup with type hints.

also need to add docstring for register file
